example/simple_pipeline/odp_simple_pipeline.c

```diff
@@ -0,0 +1,938 @@
+	int pkts, events, sent, drops;
+
+	odp_barrier_wait(&global->init_barrier);
+
+	while (!global->exit_threads) {
+		pkts = odp_pktin_recv(pktin_queue, pkt_tbl, MAX_PKT_BURST);
+		if (odp_unlikely(pkts <= 0))
+			continue;
+
+		stats->s.rx_cnt += pkts;
+
+		events = prep_events(pkt_tbl, event_tbl, pkts);
+		drops = events - pkts;
+		if (odp_unlikely(drops))
+			stats->s.rx_drops += pkts - events;
+
+		sent = odp_queue_enq_multi(out_queue, event_tbl, events);
+		if (odp_unlikely(sent < 0))
+			sent = 0;
+
+		stats->s.tx_cnt += sent;
+
+		drops = events - sent;
+		if (odp_unlikely(drops)) {
+			stats->s.tx_drops += drops;
+			odp_packet_free_multi(&pkt_tbl[sent], drops);
+		}
+	}
+
+	/* Wait until pktio devices are stopped */
+	odp_barrier_wait(&global->term_barrier);
+
+	return 0;
+}
+
+static inline int tx_thread(void *arg)
+{
+	thread_args_t *thr_args = arg;
+	odp_event_t event_tbl[MAX_PKT_BURST];
+	odp_packet_t pkt_tbl[MAX_PKT_BURST];
+	odp_queue_t rx_queue = thr_args->rx_queue;
+	odp_pktout_queue_t pktout_queue = global->if1out;
+	stats_t *stats = &thr_args->stats;
+	int events, sent, tx_drops;
+
+	odp_barrier_wait(&global->init_barrier);
+
+	while (!global->exit_threads) {
+		events = odp_queue_deq_multi(rx_queue, event_tbl,
+					     MAX_PKT_BURST);
+		if (odp_unlikely(events <= 0))
+			continue;
+
+		stats->s.rx_cnt += events;
+
+		odp_packet_from_event_multi(pkt_tbl, event_tbl, events);
+
+		sent = odp_pktout_send(pktout_queue, pkt_tbl, events);
+		if (odp_unlikely(sent < 0))
+			sent = 0;
+
+		stats->s.tx_cnt += sent;
+
+		tx_drops = events - sent;
+		if (odp_unlikely(tx_drops)) {
+			stats->s.tx_drops += tx_drops;
+			odp_packet_free_multi(&pkt_tbl[sent], tx_drops);
+		}
+	}
+
+	/* Wait until pktio devices are stopped */
+	odp_barrier_wait(&global->term_barrier);
+
+	/* Empty queue before exiting */
+	events = 1;
+	while (events > 0) {
+		events = odp_queue_deq_multi(rx_queue, event_tbl,
+					     MAX_PKT_BURST);
+
+		if (events > 0)
+			odp_event_free_multi(event_tbl, events);
+	}
+
+	return 0;
+}
+
+/*
+ * Work on packets
+ */
+static inline void work_on_events(odp_event_t event_tbl[], unsigned int num)
+{
+	unsigned int i;
+
+	for (i = 0; i < num; i++) {
+		odp_packet_t pkt = odp_packet_from_event(event_tbl[i]);
+
+		if (odp_hash_crc32c(odp_packet_data(pkt),
+				    odp_packet_seg_len(pkt), 123) == DUMMY_HASH)
+			printf("Dummy hash match\n");
+	}
+}
+
+static inline int worker_thread(void *arg ODP_UNUSED)
+{
+	thread_args_t *thr_args = arg;
+	odp_event_t event_tbl[MAX_PKT_BURST];
+	stats_t *stats = &thr_args->stats;
+	odp_queue_t rx_queue = thr_args->rx_queue;
+	odp_queue_t tx_queue = thr_args->tx_queue;
+	int events, sent, tx_drops;
+
+	odp_barrier_wait(&global->init_barrier);
+
+	while (!global->exit_threads) {
+		events = odp_queue_deq_multi(rx_queue, event_tbl,
+					     MAX_PKT_BURST);
+
+		if (odp_unlikely(events <= 0))
+			continue;
+
+		stats->s.rx_cnt += events;
+
+		work_on_events(event_tbl, events);
```

Comment:
This could be enabled/disabled with an option. CRC cycles get dominant with large packet sizes.